wakuv2/filter_manager.go

```diff
@@ -0,0 +1,360 @@
+	FilterEventSubscribeResult   int = 3
+	FilterEventUnsubscribeResult int = 4
+	FilterEventGetStats          int = 5
+)
+
+const pingTimeout = 10 * time.Second
+
+type FilterSubs map[string]subscription.SubscriptionSet
+
+type FilterEvent struct {
+	eventType int
+	filterID  string
+	result    bool
+	peerID    peer.ID
+	tempID    string
+	sub       *subscription.SubscriptionDetails
+	ch        chan FilterSubs
+}
+
+// Methods on FilterManager maintain filter peer health
+//
+// runFilterLoop is the main event loop
+//
+// Filter Install/Uninstall events are pusehd onto eventChan
+// Subscribe, UnsubscribeWithSubscription, IsSubscriptionAlive calls
+// are invoked from goroutines and request results pushed onto eventChan
+//
+// filterSubs is the map of filter IDs to subscriptions
+
+type FilterManager struct {
+	ctx              context.Context
+	filterSubs       FilterSubs
+	eventChan        chan (FilterEvent)
+	isFilterSubAlive func(sub *subscription.SubscriptionDetails) error
+	getFilter        func(string) *common.Filter
+	onNewEnvelopes   func(env *protocol.Envelope) error
+	peers            []peer.ID
+	logger           *zap.Logger
+	settings         settings
+	node             *node.WakuNode
+}
+
+func newFilterManager(ctx context.Context, logger *zap.Logger, getFilterFn func(string) *common.Filter, settings settings, onNewEnvelopes func(env *protocol.Envelope) error, node *node.WakuNode) *FilterManager {
+	// This fn is being mocked in test
+	mgr := new(FilterManager)
+	mgr.ctx = ctx
+	mgr.logger = logger
+	mgr.getFilter = getFilterFn
+	mgr.onNewEnvelopes = onNewEnvelopes
+	mgr.filterSubs = make(FilterSubs)
+	mgr.eventChan = make(chan FilterEvent)
+	mgr.peers = make([]peer.ID, 0)
+	mgr.settings = settings
+	mgr.node = node
+	mgr.isFilterSubAlive = func(sub *subscription.SubscriptionDetails) error {
+		ctx, cancel := context.WithTimeout(ctx, pingTimeout)
+		defer cancel()
+		return mgr.node.FilterLightnode().IsSubscriptionAlive(ctx, sub)
+	}
+
+	return mgr
+}
+
+func (mgr *FilterManager) runFilterLoop(wg *sync.WaitGroup) {
+	defer wg.Done()
+	// Use it to ping filter peer(s) periodically
+	ticker := time.NewTicker(5 * time.Second)
+	defer ticker.Stop()
+
+	// Populate filter peers initially
+	mgr.peers = mgr.findFilterPeers() // ordered list of peers to select from
+
+	for {
+		select {
+		case <-mgr.ctx.Done():
+			return
+		case <-ticker.C:
+			mgr.peers = mgr.findFilterPeers()
+			mgr.pingPeers()
+		case ev := <-mgr.eventChan:
+			switch ev.eventType {
+
+			case FilterEventAdded:
+				mgr.filterSubs[ev.filterID] = make(subscription.SubscriptionSet)
+				mgr.resubscribe(ev.filterID)
+
+			case FilterEventRemoved:
+				for _, sub := range mgr.filterSubs[ev.filterID] {
+					if sub == nil {
+						// Skip temp subs
+						continue
+					}
+					go mgr.unsubscribeFromFilter(ev.filterID, sub)
+				}
+				delete(mgr.filterSubs, ev.filterID)
+
+			case FilterEventPingResult:
+				if ev.result {
+					break
+				}
+				if ev.filterID != "" {
```

Comment:
it's not immediately clear why a `filterID` being set indicates that a filter has not peers, maybe we could elaborate with a comment?, or rather the opposite, why `filterID` would be empty?

Reply:
Added an extra comment.